Tasks/PublishTestResults/publishresultsthroughexe.ts

```diff
@@ -0,0 +1,112 @@
+        testRunTitle: string, publishRunAttachments: string, testRunner: string) {
+        
+        this.matchingTestResultsFiles = matchingTestResultsFiles.slice(0);
+        this.mergeResults = mergeResults;
+        this.platform = platform;
+        this.config = config;
+        this.testRunTitle = testRunTitle;
+        this.publishRunAttachments = publishRunAttachments;
+        this.testRunner = testRunner;
+    }
+
+    public async publishResultsThroughExe(): Promise<number> {
+        let testResultsPublisherTool: tr.ToolRunner = tl.tool(this.getTestResultsPublisherLocation());
+        let envVars: { [key: string]: string; } = this.getEnvironmentVariables();
+        let args: string[] = this.getArguments();
+        testResultsPublisherTool.arg(args);
+        
+        let exitCode: number = await testResultsPublisherTool.exec(<tr.IExecOptions>{ env: envVars, ignoreReturnCode: true });
+
+        return exitCode;
+    }
+
+    private getTestResultsPublisherLocation(): string {
+        return path.join(__dirname, 'modules/TestResultsPublisher.exe');
+    }
+
+    private getArguments(): string[] {
+        let responseFilePath = this.createResponseFile();
+        // Adding '@' because this is a response file argument
+        let args = ['@' + responseFilePath];
+        return args;
+
+    }
+
+    private createResponseFile(): string {
+        let responseFilePath: string = path.join(__dirname, 'tempResponseFile.txt');
+
+        // Adding quotes around matching file names
+        this.modifyMatchingFileName();
+        
+        // Preparing File content
+        let fileContent: string = os.EOL + this.matchingTestResultsFiles.join(os.EOL);
+
+        // Writing matching file names in the response file
+        try {
+            fs.writeFileSync(responseFilePath, fileContent);
+        }
+        catch (ex) {
+            tl.warning("Exception while writing to response file: " + ex);
+        }
+
+        return responseFilePath;
+    }
+
+    private modifyMatchingFileName(): void {
+        for (let i = 0; i < this.matchingTestResultsFiles.length; i++) {
+            // We need to add quotes around the file name because the file name can contain spaces.
+            // The quotes will be handled by response file reader.
+            this.matchingTestResultsFiles[i] = '\"' + this.matchingTestResultsFiles[i] + '\"';
+        }
+    }
+
+    private getEnvironmentVariables(): { [key: string]: string; } {
+        let envVars: { [key: string]: string; } = {};
+
+        envVars = this.addToProcessEnvVars(envVars, 'collectionurl', tl.getVariable('System.TeamFoundationCollectionUri'));
+        envVars = this.addToProcessEnvVars(envVars, 'accesstoken', tl.getEndpointAuthorizationParameter('SystemVssConnection', 'AccessToken', false));
+
+        envVars = this.addToProcessEnvVars(envVars, 'testrunner', this.testRunner);
+        envVars = this.addToProcessEnvVars(envVars, 'mergeresults', this.mergeResults);
```

Comment:
why some of these should be passed in constructor while some of them are called using tl.getVariable?

Write a simple model class and get all the required inputs there and pass that model to Publisher Class.

Reply:
Some of them were already present used by Command. I'm getting the variables which were not being used by Command here by tl.getVariable